app/services/products_renderer.rb

```diff
@@ -51,10 +51,56 @@ def enterprise_fee_calculator
     OpenFoodNetwork::EnterpriseFeeCalculator.new distributor, order_cycle
   end
 
-  def filter_and_paginate(query)
-    results = query.ransack(args[:q]).result
+  # TODO refactor this, distributed_products should be able to give use the relation based
+  # on the sorting method, same for ordering. It would prevent the SQL implementation from
+  # leaking here
+  def products_relation
+    if distributor.preferred_shopfront_product_sorting_method == "by_category" &&
+       distributor.preferred_shopfront_taxon_order.present?
+      return distributed_products.products_taxons_relation
+    end
+
+    distributed_products.products_supplier_relation
+  end
+
+  def filter(query)
+    supplier_properties = args[:q]&.slice("with_variants_supplier_properties")
+
+    ransack_results = query.ransack(args[:q]).result.to_a
+
+    return ransack_results if supplier_properties.blank?
+
+    with_properties = args[:q]&.dig("with_properties")
+    supplier_properties_results = []
+
+    if supplier_properties.present?
+      # We can't search on an association's scope with ransack, a work around is to define
+      # the a scope on the parent (Spree::Product) but because we are joining on "first_variant"
+      # to get the supplier it doesn't work, so we do the filtering manually here
+      # see:
+      #   OrderCycleDistributedProducts#products_supplier_relation
+      #   OrderCycleDistributedProducts#supplier_property_join
+      supplier_property_ids = supplier_properties["with_variants_supplier_properties"]
+      supplier_properties_results = distributed_products.supplier_property_join(query).
+        where(producer_properties: { property_id: supplier_property_ids }).
+        where(inherits_properties: true)
+    end
+
+    if supplier_properties_results.present? && with_properties.present?
+      # apply "OR" between property search
+      return ransack_results | supplier_properties_results
+    end
+
+    # Intersect the result to apply "AND" with other search criteria
+    return ransack_results.intersection(supplier_properties_results) \
+      unless supplier_properties_results.empty?
+
+    # We should get here but just in case we return the ransack results
+    ransack_results
+  end
 
```

Comment:
This works but looks kinda nasty, I want to fix this by moving some of the logic in `app/services/order_cycles/distributed_products_service.rb` but I'll do that in a separate PR.